### PR TITLE
Migrate to hatch

### LINKED ISSUE #747: Do an audit of requirements files, pyproject.toml, ci config, tox
We seem to have added some complexity and it probably makes sense to audit these workflows.

### PR DESCRIPTION
**Pull request by willingc**
Closes #747

### CODE REVIEW

**Comment by MSeal** [comment]
Can we get a draft version without the linting / quoting changes? Would be a lot easier to review / iterate if we can undo that part for now.

**Comment by willingc** [comment]
Absolutely @MSeal. I would love to see the precommit bot to avoid autofixing things.

Right now this is a draft. I want to see if by going to more modern packaging if we can get rid of tox and simplify CI.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -21,31 +20,21 @@ repos:
...
       - id: detect-private-key
 
+  - repo: https://github.com/charliermarsh/ruff-pre-commit
+    rev: v0.1.2
+    hooks:
+    -   id: ruff
+        args:
+        - --fix
+    -   id: ruff-format
+
```
Comment (Borda) [inline .pre-commit-config.yaml:30]:
```suggestion
```
already added with switch flake8 -> ruff #752

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -21,31 +20,21 @@ repos:
...
-  #  rev: 5.10.1
-  #  hooks:
-  #    - id: isort
-
-  - repo: https://github.com/psf/black
-    rev: 23.11.0
-    hooks:
-      - id: black
-        name: Black code
-        args: ["-S"]
```
Comment (Borda) [inline .pre-commit-config.yaml:47]: why drop Black?

File: docs/conf.py [hunk 3]
```diff
@@ -80,7 +80,7 @@
 exclude_patterns = ['_build', 'Thumbs.db', '.DS_Store', 'UPDATE.md']
 
 # The name of the Pygments (syntax highlighting) style to use.
-pygments_style = "sphinx"
+pygments_style = 'sphinx'
```
Comment (Borda) [inline docs/conf.py:83]: lets fo these changes in separate PR so it is easier to follow what is the main change/focus of this PR

File: pyproject.toml [hunk 4]
```diff
@@ -1,30 +1,192 @@
...
+line-length = 120
+indent-width = 4
+select = [
+    "E",  # pycodestyle errors
+    "W",  # pycodestyle warnings
+    "F",  # pyflakes
+    "I",  # isort
+    "C",  # flake8-comprehensions
+    "B",  # flake8-bugbear
+    "UP",  # pyupgrade
```
Comment (Borda) [inline pyproject.toml:151]: lets rather apply these one by one in separate PRs

File: pyproject.toml [hunk 5]
```diff
@@ -1,30 +1,192 @@
...
+    "ipython>=5.0",
+    "ipywidgets",
+    "moto",
+    "notebook",
+    "pip>=18.1",
+    "pre-commit",
+    "pytest-cov>=2.6.1",
+    "pytest-env>=0.6.2",
+    "pytest-mock>=1.10",
+    "pytest>=4.1",
```
Comment (Borda) [inline pyproject.toml:98]: seems like a mix of develop and base requirement, I doubt pre-commit is required for this Project to run :)

File: pyproject.toml [hunk 6]
```diff
@@ -1,30 +1,192 @@
...
+    "coverage",
+    "flake8",
+    "gcsfs>=0.2.0",
+    "google_compute_engine",
+    "ipython>=5.0",
+    "ipywidgets",
+    "moto",
+    "notebook",
+    "pip>=18.1",
+    "pre-commit",
```
Comment (Borda) [inline pyproject.toml:94]: does it have also split to dev and base requirements?

**Comment by willingc** [comment]
@Borda I'm going to close this PR but am leaving open right now as a WIP as I'm going to write a tutorial for the hatch maintainer about migrating an existing repo to hatch.

**Comment by Borda** [comment]
> I'm going to write a tutorial for the hatch maintainer about migrating an existing repo to hatch.

looking forward to reading it it would be an interesting solution for a few of my projects...

File: .pre-commit-config.yaml (hunk 2 above)
Comment (akx) [reply .pre-commit-config.yaml:47]: Because this adds ruff-format instead.

**Comment by akx** [comment]
@Borda FWIW, I've found that [`hatch new --init`](https://hatch.pypa.io/latest/intro/#existing-project) does a pretty good job in many cases; you can combine that with [`flake8-to-ruff`](https://pypi.org/project/flake8-to-ruff/), [`pyproject-migrator`](https://pypi.org/project/pyproject-migrator/) and [`pylint-to-ruff`](https://pypi.org/project/pylint-to-ruff/) in a suitable combination... 😁

**Pull request closed without merging**